File: templates/problems/04_stack/07_largest_rect_histogram/largest_rect_histogram_test.go
package largest_rect_histogram

import "testing"

func TestLargestRectangleArea(t *testing.T) {
	tests := []struct {
		name    string
		heights []int
		want    int
	}{
		{"example 1", []int{2, 1, 5, 6, 2, 3}, 10},
		{"single bar", []int{5}, 5},
		{"increasing", []int{1, 2, 3, 4, 5}, 9},
		{"decreasing", []int{5, 4, 3, 2, 1}, 9},
		{"all same", []int{3, 3, 3, 3}, 12},
		{"two bars", []int{2, 4}, 4},
		{"valley", []int{6, 2, 5, 4, 5, 1, 6}, 12},
		{"empty", []int{}, 0},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := largestRectangleArea(tt.heights); got != tt.want {
				t.Errorf("largestRectangleArea(%v) = %d, want %d", tt.heights, got, tt.want)
			}
		})
	}
}

File: templates/problems/04_stack/07_largest_rect_histogram/typescript/largestRectHistogram.test.ts
import { describe, it, expect } from 'vitest';
import { largestRectangleArea } from './largestRectHistogram';

describe('LargestRectHistogram', () => {
  it('example 1', () => {
    expect(largestRectangleArea([2, 1, 5, 6, 2, 3])).toBe(10);
  });

  it('single bar', () => {
    expect(largestRectangleArea([5])).toBe(5);
  });

  it('increasing', () => {
    expect(largestRectangleArea([1, 2, 3, 4, 5])).toBe(9);
  });

  it('decreasing', () => {
    expect(largestRectangleArea([5, 4, 3, 2, 1])).toBe(9);
  });

  it('all same', () => {
    expect(largestRectangleArea([3, 3, 3, 3])).toBe(12);
  });

  it('two bars', () => {
    expect(largestRectangleArea([2, 4])).toBe(4);
  });

  it('valley', () => {
    expect(largestRectangleArea([6, 2, 5, 4, 5, 1, 6])).toBe(12);
  });

  it('empty', () => {
    expect(largestRectangleArea([])).toBe(0);
  });
});
